feeder/database/models.py

```diff
@@ -299,10 +303,206 @@ async def clear_for_device(cls, device_id):
+            values["g_per_tbsp"] = g_per_tbsp
+        if tbsp_per_feeding:
+            values["tbsp_per_feeding"] = tbsp_per_feeding
+        if budget_tbsp:
+            values["budget_tbsp"] = budget_tbsp
+        query = recipes.update().where(recipes.c.id == recipe_id).values(
+            **values
+        )
+        return await db.execute(query)
+
+
+# We will store a level reference every time someone manually updates the amount
+# of food in the hopper. From there we can find the latest reference and subtract
+# the total volume of all feed results since that reference. That will give us a
+# estimate of how much food remains.
+hopper_level_references = Table(
+    "hopper_references",
     metadata,
     Column("device_hid", Text(), ForeignKey("kronos_device.hid"), primary_key=True),
+    Column("timestamp", Integer(), primary_key=True),
+    Column("level", Integer(), nullable=False)
+)
+
+
+class HopperLevelRef:
+    @classmethod
+    async def get(cls, device_id):
+        # Kenny thinks it is a 40 cup hopper based on measurements... I'm not about to cover myself
+        # in cat food to prove him right or wrong.
+        max_hopper_cups = 40
+
+        device_results = await KronosDevices.get(device_hid=device_id)
+        device = device_results[0]
+        latest_ref_query = hopper_level_references.select().order_by(
+            desc(hopper_level_references.c.timestamp)
+        )
+        latest_ref = await db.fetch_one(latest_ref_query)
+        if not latest_ref:
+            raise HTTPException(404, detail=f"Hopper level not set for {device_id}")
+
+        logger.debug("Hopper level last set to %d on %d", latest_ref.level, latest_ref.timestamp)
+
+        dispensed_grams_query = select(
+            [func.sum(feeding_event.c.grams_expected)]
+        ).select_from(feeding_event).where(
+            feeding_event.c.start_time >= latest_ref.timestamp
+        ).where(feeding_event.c.device_hid == device_id)
+        # TODO: So, this isn't _technically_ the most correct way to do this.
+        # We are making the assumption that all of the dispensed feeds from the last reference
+        # are the current recipe. While this isn't an unfair assumption, since food level
+        # would change and should be updated by the user if the food changes, we are relying on
+        # the user to do the right thing... Which, yah, good luck with that.
+        dispensed_grams = await db.fetch_val(dispensed_grams_query)
+        if not dispensed_grams:
+            dispensed_grams = 0
+        logger.debug("%d grams of food have been dispensed since %d", dispensed_grams, latest_ref.timestamp)
+        recipe_query = recipes.select().where(recipes.c.id == device.currentRecipe)
+        recipe = await db.fetch_one(recipe_query)
+        if not recipe:
+            raise HTTPException(400, detail="No recipe set for device, cannot calculate hopper level!")
+        dispensed_cups = (dispensed_grams / recipe.g_per_tbsp) / 16
```

Comment:
Added the variable. I have no idea, but the fact that it would require different hardware for them to change the minimum dispensed amount any smaller than 1 tbsp and we are weighing things in grams makes me think it isn't going to be an issue. On a side note though, I did add the option when adding a pet to define their weight in kg and regardless of unit, we are converting and storing weight as grams.